server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -72,23 +72,47 @@ public class DesiredBalanceComputer {
 
     private TimeValue progressLogInterval;
     private long maxBalanceComputationTimeDuringIndexCreationMillis;
+    private long numComputeCallsSinceLastConverged;
+    private long numIterationsSinceLastConverged;
+    private long lastConvergedTimeMillis;
+    private long lastNotConvergedLogMessageTimeMillis;
 
     public DesiredBalanceComputer(ClusterSettings clusterSettings, TimeProvider timeProvider, ShardsAllocator delegateAllocator) {
+        this(clusterSettings, timeProvider, delegateAllocator, STATIC_LOGGER);
+    }
+
+    // Package access for testing.
+    DesiredBalanceComputer(ClusterSettings clusterSettings, TimeProvider timeProvider, ShardsAllocator delegateAllocator, Logger logger) {
+        this.logger = logger;
         this.delegateAllocator = delegateAllocator;
         this.timeProvider = timeProvider;
+        this.numComputeCallsSinceLastConverged = 0;
+        this.numIterationsSinceLastConverged = 0;
+        this.lastConvergedTimeMillis = timeProvider.relativeTimeInMillis();
+        this.lastNotConvergedLogMessageTimeMillis = lastConvergedTimeMillis;
         clusterSettings.initializeAndWatch(PROGRESS_LOG_INTERVAL_SETTING, value -> this.progressLogInterval = value);
         clusterSettings.initializeAndWatch(
             MAX_BALANCE_COMPUTATION_TIME_DURING_INDEX_CREATION_SETTING,
             value -> this.maxBalanceComputationTimeDuringIndexCreationMillis = value.millis()
         );
     }
 
+    public record LastConvergenceInfo(long numComputeCallsSince, long numIterationsSince, long timestampMillis) {}
```

Comment:
I'd rather not add another class just for exposing these values in tests. Let's have the tests access these values directly as needed instead. You can add a test utility for asserting that they're all as expected at once instead if you'd like.

Also the tests are in the same package as the production code so the accessors can be package-private.